File: TransformListener.lua
local ffi = require 'ffi'
local torch = require 'torch'
local ros = require 'ros.env'
local utils = require 'ros.utils'
local std = ros.std
local tf = ros.tf

local TransformListener = torch.class('tf.TransformListener', tf)

local f

function init()
  local TransformListener_method_names = {
    "new",
    "delete",
    "clear",
    "getFrameStrings",
    "lookupTransform",
    "waitForTransform",
    "canTransform",
    "lookupTransformFull",
    "waitForTransformFull",
    "canTransformFull",
    "resolve",
    "getLatestCommonTime",
    "chainAsVector",
    "getParent",
    "frameExists",
    "getCacheLength",
    "getTFPrefix"
  }
  
  f = utils.create_method_table("tf_TransformListener_", TransformListener_method_names)
end

init()

function TransformListener:__init()
  self.o = f.new()
end

function TransformListener:clear()
  f.clear(self.o)
end

function TransformListener:getFrameStrings()
  local result = std.StringVector()
  f.getFrameStrings(self.o, result:cdata())
  return result
end

function TransformListener:lookupTransform(target_frame, source_frame, time, result)
  result = result or tf.StampedTransform()
  f.lookupTransform(self.o, target_frame, source_frame, time:cdata(), result:cdata())
  return result
end

function TransformListener:waitForTransform(target_frame, source_frame, time, timeout, may_throw)
  timeout = timeout or ros.Duration(10)
  if may_throw then
    local error_msg = std.String()
    if not f.waitForTransform(self.o, target_frame, source_frame, time:cdata(), timeout:cdata(), error_msg:cdata()) then
      error(error_msg:get())
    end
    return true
  else
    return f.waitForTransform(self.o, target_frame, source_frame, time:cdata(), timeout:cdata(), ffi.NULL)
  end
end

function TransformListener:canTransform(target_frame, source_frame, time)
  return f.canTransform(self.o, target_frame, source_frame, time:cdata())
end

function TransformListener:lookupTransformFull(target_frame, target_time, source_frame, source_time, fixed_frame, result)
  result = result or tf.StampedTransform()
  f.lookupTransformFull(self.o, target_frame, target_time:cdata(), source_frame, source_time:cdata(), fixed_frame, result:cdata())
  return result
end

function TransformListener:waitForTransformFull(target_frame, target_time, source_frame, source_time, fixed_frame, timeout, may_throw)
  timeout = timeout or ros.Duration(10)
  if may_throw then
    local error_msg = std.String()
    if not f.waitForTransformFull(self.o, target_frame, target_time:cdata(), source_frame, source_time:cdata(), fixed_frame, timeout:cdata(), error_msg:cdata()) then
      error(error_msg:get())
    end
    return true
  else
    return f.waitForTransformFull(self.o, target_frame, target_time:cdata(), source_frame, source_time:cdata(), fixed_frame, timeout:cdata(), ffi.NULL)
  end
end

function TransformListener:canTransformFull(target_frame, target_time, source_frame, source_time, fixed_frame)
  return f.canTransformFull(self.o, target_frame, target_time:cdata(), source_frame, source_time:cdata(), fixed_frame)
end

function TransformListener:resolve(frame_name)
  local name = std.String()
  f.resolve(self.o, frame_name, name:cdata())
  return name:get()
end

function TransformListener:getLatestCommonTime(source_frame, target_frame)
  local result = ros.Time()
  local error_msg = std.String()
  if t.getLatestCommonTime(self.o, source_frame, target_frame, result:cdata(), error_msg:cdata()) == 0 then
    error(error_msg:get())
  end
end

function TransformListener:chainAsVector(target_frame, target_time, source_frame, source_time, fixed_frame, result)
  result = result or std.StringVector
  t.chainAsVector(self.o, target_frame, target_time:cdata(), source_frame, source_time:cdata(), fixed_frame, result:cdata())
  return result
end

function TransformListener:getParent(frame_id, time)
  time = time or ros.Time(0)
  local parent = std.String()
  t.getParent(self.o, frame_id, time:cdata(), parent:cdata())
  return parent:get()
end

function TransformListener:frameExists(frame_id)
  return t.frameExists(self.o, frame_id)
end

function TransformListener:getCacheLength()
  local duration = ros.Duration()
  t.getCacheLength(self.o, duration:cdata())
  return duration
end

function TransformListener:getTFPrefix()
  local prefix = std.String()
  t.getTFPrefix(self.o, prefix:cdata())
  return prefix
end

File: src/tf/transform_listener.cpp
#include "torch-tf.h"
#include <tf/transform_listener.h>

TFIMP(tf::TransformListener *, TransformListener, new)() {
  return new tf::TransformListener();
}

TFIMP(void, TransformListener, delete)(tf::TransformListener *self) {
  delete self;
}

TFIMP(void, TransformListener, clear)(tf::TransformListener *self) {
  self->clear();
}

TFIMP(void, TransformListener, getFrameStrings)(tf::TransformListener *self, std::vector<std::string> *result) {
  try
    {
      self->getFrameStrings(*result);
    }
  catch (std::runtime_error& e)
    {
      ROS_ERROR("Exception: [%s]", e.what());
    }
}

TFIMP(void, TransformListener, lookupTransform)(
  tf::TransformListener *self,
  const char *target_frame,
  const char *source_frame, ros::Time *time,
  tf::StampedTransform *result
) {
  try
    {
      self->lookupTransform(target_frame, source_frame, *time, *result);
    }
  catch (std::runtime_error& e)
    {
      ROS_ERROR("Exception: [%s]", e.what());
    }
}

TFIMP(bool, TransformListener, waitForTransform)(
  tf::TransformListener *self,
  const char *target_frame,
  const char *source_frame,
  ros::Time *time,
  ros::Duration *timeout,
  std::string *error_msg
) {
  return self->waitForTransform(target_frame, source_frame, *time, *timeout, ros::Duration(0.01), error_msg);
}

TFIMP(bool, TransformListener, canTransform)(
  tf::TransformListener *self,
  const char *target_frame,
  const char *source_frame,
  ros::Time *time
) {
  return self->canTransform(target_frame, source_frame, *time, NULL);
}

TFIMP(void, TransformListener, lookupTransformFull)(tf::TransformListener *self,
  const char *target_frame, ros::Time *target_time,
  const char *source_frame, ros::Time *source_time,
  const char *fixed_frame, tf::StampedTransform *result
  ) {
  self->lookupTransform(target_frame, *target_time, source_frame, *source_time, fixed_frame, *result);
}

TFIMP(bool, TransformListener, waitForTransformFull)(tf::TransformListener *self,
  const char *target_frame, ros::Time *target_time,
  const char *source_frame, ros::Time *source_time,
  const char *fixed_frame, ros::Duration *timeout, std::string *error_msg
) {
  return self->waitForTransform(target_frame, *target_time, source_frame, *source_time, fixed_frame, *timeout, ros::Duration(0.01), error_msg);
}

TFIMP(bool, TransformListener, canTransformFull)(tf::TransformListener *self,
  const char *target_frame, ros::Time *target_time,
  const char *source_frame, ros::Time *source_time,
  const char *fixed_frame
) {
  return self->canTransform(target_frame, *target_time, source_frame, *source_time, fixed_frame, NULL);
}

TFIMP(void, TransformListener, resolve)(tf::TransformListener *self, const char *frame_name, std::string *result) {
  *result = self->resolve(frame_name);
}

TFIMP(int, TransformListener, getLatestCommonTime)(
  tf::TransformListener *self,
  const char *source_frame,
  const char *target_frame,
  ros::Time *time,
  std::string *error_string
) {
  return self->getLatestCommonTime(source_frame, target_frame, *time, error_string);
}

TFIMP(void, TransformListener, chainAsVector)(tf::TransformListener *self,
  const char *target_frame, ros::Time *target_time,
  const char *source_frame, ros::Time *source_time,
  const char *fixed_frame, std::vector<std::string> *result
) {
  self->chainAsVector(target_frame, *target_time, source_frame, *source_time, fixed_frame, *result);
}

TFIMP(bool, TransformListener, getParent)(
  tf::TransformListener *self,
  const char* frame_id,
  ros::Time *time,
  std::string *result
) {
  return self->getParent(frame_id, *time, *result);
}

TFIMP(bool, TransformListener, frameExists)(tf::TransformListener *self, const char *frame_id) {
  try
    {
      return self->frameExists(frame_id);
    }
  catch (std::runtime_error& e)
    {
      ROS_ERROR("Exception: [%s]", e.what());
      return false;
    }
}

TFIMP(void, TransformListener, getCacheLength)(tf::TransformListener *self, ros::Duration *result) {
  *result = self->getCacheLength();
}

TFIMP(void, TransformListener, getTFPrefix)(tf::TransformListener *self, std::string *result) {
  *result = self->getTFPrefix();
}
